### PR TITLE
[MPS] Fix CPU scalar storage_offset ignored in binary op fast path

### LINKED ISSUE #187117: [MPS] Binary ops with a CPU 0-dim view operand ignore storage_offset — silently read element 0
### 🐛 Describe the bug

When a binary op mixes an MPS tensor with a **CPU 0-dim tensor that is a view with a nonzero `storage_offset()`** (e.g. `s[4]` of a 1-D CPU tensor), the MPS backend silently uses the **wrong scalar value** instead of raising a device-mismatch error or computing the correct result. The result matches what you'd get with **element 0** of the CPU tensor's storage, bit-exactly — i.e. the scalar-promotion path appears to read the operand's storage base pointer and ignore `storage_offset`.

```python
import torch

torch.manual_seed(0)
a = torch.randn(3, 4)
s = torch.rand(8) + 0.1   # CPU 1-D tensor

# s[4] is a CPU 0-dim view with storage_offset()==4
diff = ((a / s[4]) - (a.to('mps') / s[4]).cpu()).abs().max()
print(diff)  # tensor(3.4439)  <- should be 0

# The MPS result is bit-exact with using s[0] instead of s[4]:
print(((a / s[0]) - (a.to('mps') / s[4]).cpu()).abs().max())  # tensor(0.)
```

### Scope (verified on 2.12.0)

Affected:
- `div`, `mul`, `add`, `sub` (presumably the whole `*_dense_scalar_*` binary-kernel family)
- Out-of-place and in-place (`a_mps.div_(s[4])`)
- Any nonzero `storage_offset` (tested 1, 4, 7 — all read element 0; offset 0 is correct by coincidence)
- Strided views too (`s[::2][2]`, offset 4 — same wrong result)
- Multiple dtypes: `float32`, `float16`, `int64`

NOT affected (correct results):
- A plain CPU 0-dim tensor with offset 0, e.g. `torch.tensor(2.0)`
- An **MPS-resident** 0-dim view (`s.to('mps')[4]`) — that path gathers correctly
- CPU↔CPU, the same expression on CPU is the correct reference

Workaround: `.clone()` (or `.item()`) the 0-dim operand before the op.

### Root-cause hypothesis

On the mixed-device fast path that promotes a CPU 0-dim tensor operand to a kernel scalar argument (the `*_dense_scalar_*` Metal kernels), the scalar bytes appear to be taken from the tensor's storage **base** pointer rather than `storage_offset()` into it, so any indexed/sliced 0-dim view reads element 0. This is the arithmetic-op analog of #182822 (bitwise ops on bool ignoring `storage_offset` of a 0-d operand), which was fixed for bitwise ops only by migrating them to the TensorIterator infrastructure (#182839). The same `*_dense_scalar_*` path has had other device-mixing issues: #186580, #181650.

Note this is `module: correctness (silent)` bait: `mps_tensor * cpu_tensor[i]` is a very natural way to write per-channel/per-index normalization (e.g. `x / std[k]`), it raises no error, and it only corrupts results when `i != 0`.

### Versions

```
PyTorch: 2.12.0 (pip, arm64)
macOS: 15.6 (Darwin 24.6.0)
Chip: Apple M4
Python: 3.11.15
torch.backends.mps.is_available(): True
```

cc @malfet @aditvenk @kulinseth @DenisVieriu97 @jhavukainen @Isalia20

### PR DESCRIPTION
**Pull request by senekis316**
## Summary

Fixes #187117.

`mtl_setBuffer` in `OperationUtils.h` was calling `[encoder setBytes:t.storage().data() ...]` for CPU 0-dim tensor scalars. This passes the raw storage base address, silently ignoring `storage_offset`. A 0-dim view like `s[4]` has `storage_offset=4`, so the Metal kernel was reading element 0 of the backing storage instead of element 4.

**Fix:** use `t.const_data_ptr()` instead, which already incorporates the offset (`storage().data() + storage_offset * element_size`). The double/complex-double branches above it already use `const_data_ptr()` correctly — this brings the fallback path in line.

```python
# Before fix
s = torch.tensor([0.0, 1.0, 2.0, 3.0, 4.0])
a = torch.ones(3, device='mps')
print(a + s[4])  # tensor([1., 1., 1.])  ← wrong, read s[0]

# After fix
print(a + s[4])  # tensor([5., 5., 5.])  ← correct
```

## Test plan

- Added `test_cpu_scalar_storage_offset_binary` to `test/test_mps.py` covering float and integer dtypes, both lhs and rhs scalar positions, for all offsets 0–4.
- Verified the test **fails** on the unfixed code and **passes** after the fix.
- Ran `pytest test/test_mps.py -k "binary or scalar or storage_offset"`: 165 passed, 0 failed.

### CODE REVIEW

**Comment by malfet** [comment]
@claude review this

**Review by Skylion007 — APPROVED** [review]
